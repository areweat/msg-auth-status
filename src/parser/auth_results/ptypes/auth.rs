//! auth ptype and it's properties
//!
//! See IANA Assignments
//! https://www.iana.org/assignments/email-auth/email-auth.xhtml
//!
//! And Auth in RFC 8601 s. 2.7

#[derive(Debug)]
pub enum AuthProperty<'hdr> {
    SmtpAuth(&'hdr str),
    MailFrom(&'hdr str),
}

#[derive(Debug)]
pub enum AuthPtype {
    Smtp,
    SmtpDotAuth,
    Mail,
    MailDotAuth,
}
