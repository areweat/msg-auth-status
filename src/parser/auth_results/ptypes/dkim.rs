//! Parsing dkim property types & values

use super::*;

pub mod dkim_property_key;
pub mod dkim_property_value;
