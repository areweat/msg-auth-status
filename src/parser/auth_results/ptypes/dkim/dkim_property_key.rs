//! Parsing dkim property types & values

use crate::error::AuthResultsError;

/// IANA Email Authentication Methods ptype / property Mapping stages
#[derive(Debug, PartialEq)]
pub enum DkimHeaderPropertyKey<'hdr> {
    TagD,
    TagI,
    TagB,
    TagA,
    TagS,
    Rfc5322From,
    Unknown(&'hdr str),
}

impl<'hdr> TryFrom<DkimHeaderPropertyKeyToken<'hdr>> for DkimHeaderPropertyKey<'hdr> {
    type Error = AuthResultsError<'hdr>;
    fn try_from(token: DkimHeaderPropertyKeyToken<'hdr>) -> Result<Self, Self::Error> {
        let okk = match token {
            DkimHeaderPropertyKeyToken::TagD => Self::TagD,
            DkimHeaderPropertyKeyToken::TagI => Self::TagI,
            DkimHeaderPropertyKeyToken::TagB => Self::TagB,
            DkimHeaderPropertyKeyToken::TagA => Self::TagA,
            DkimHeaderPropertyKeyToken::TagS => Self::TagS,
            DkimHeaderPropertyKeyToken::Rfc5322From => Self::Rfc5322From,
            DkimHeaderPropertyKeyToken::Unknown(val) => Self::Unknown(val),
            _ => return Err(AuthResultsError::ParsePtypeBugInvalidProperty),
        };
        Ok(okk)
    }
}

#[derive(Debug, PartialEq)]
pub enum DkimPolicyPropertyKey<'hdr> {
    Unknown(&'hdr str),
}

impl<'hdr> TryFrom<DkimPolicyPropertyKeyToken<'hdr>> for DkimPolicyPropertyKey<'hdr> {
    type Error = AuthResultsError<'hdr>;
    fn try_from(token: DkimPolicyPropertyKeyToken<'hdr>) -> Result<Self, Self::Error> {
        let okk = match token {
            DkimPolicyPropertyKeyToken::Unknown(val) => Self::Unknown(val),
            _ => return Err(AuthResultsError::ParsePtypeBugInvalidProperty),
        };
        Ok(okk)
    }
}

//----------
// Parsing dkim property
//----------

use super::{parse_comment, CommentToken};

use logos::{Lexer, Logos};

#[derive(Debug, Logos)]
pub enum DkimHeaderPropertyKeyToken<'hdr> {
    #[token("a", priority = 1)]
    TagA,

    #[token("b", priority = 1)]
    TagB,

    #[token("d", priority = 1)]
    TagD,

    #[token("i", priority = 1)]
    TagI,

    #[token("s", priority = 1)]
    TagS,

    #[token("from", priority = 6)]
    Rfc5322From,

    #[token("(", priority = 2)]
    CommentStart,

    #[regex(r"(f[a-po-z]|[ce-hk-rt-z])[a-z\-_]*", |lex| lex.slice(), priority = 3)]
    Unknown(&'hdr str),

    #[regex(r"\s+", |lex| lex.slice(), priority = 6)]
    #[allow(dead_code)]
    WhiteSpaces(&'hdr str),
}

pub fn parse_dkim_header_property_key<'hdr>(
    lexer: &mut Lexer<'hdr, DkimHeaderPropertyKeyToken<'hdr>>,
) -> Result<DkimHeaderPropertyKey<'hdr>, AuthResultsError<'hdr>> {
    while let Some(token) = lexer.next() {
        match token {
            Ok(
                DkimHeaderPropertyKeyToken::TagD
                | DkimHeaderPropertyKeyToken::TagI
                | DkimHeaderPropertyKeyToken::TagB
                | DkimHeaderPropertyKeyToken::TagA
                | DkimHeaderPropertyKeyToken::TagS
                | DkimHeaderPropertyKeyToken::Rfc5322From,
            ) => {
                let property = token.map_err(|_| AuthResultsError::ParsePtypeBugPropertyGating)?;
                let mapped_property_res: Result<
                    DkimHeaderPropertyKey<'hdr>,
                    AuthResultsError<'hdr>,
                > = property.try_into();
                let mapped_property = mapped_property_res
                    .map_err(|_| AuthResultsError::ParsePtypeBugInvalidProperty)?;
                return Ok(mapped_property);
            }
            Ok(DkimHeaderPropertyKeyToken::WhiteSpaces(_)) => {
                // cont
            }
            Ok(DkimHeaderPropertyKeyToken::CommentStart) => {
                let mut comment_lexer = CommentToken::lexer(lexer.remainder());
                match parse_comment(&mut comment_lexer) {
                    Ok(_comment) => {}
                    Err(e) => return Err(e),
                }
                lexer.bump(comment_lexer.span().end);
            }
            _ => {
                let cut_slice = &lexer.source()[lexer.span().start..];
                let cut_span = &lexer.source()[lexer.span().start..lexer.span().end];

                let detail = crate::error::ParsingDetail {
                    component: "parse_dkim_header_property_key",
                    span_start: lexer.span().start,
                    span_end: lexer.span().end,
                    source: lexer.source(),
                    clipped_span: cut_span,
                    clipped_remaining: cut_slice,
                };
                return Err(AuthResultsError::ParsingDetailed(detail));
            }
        }
    }
    Err(AuthResultsError::RunAwayDkimPropertyKey)
}

//----------
// Parsing dkim policy property
//----------

#[derive(Debug, Logos)]
pub enum DkimPolicyPropertyKeyToken<'hdr> {
    #[token("(", priority = 1)]
    CommentStart,

    #[regex(r"[a-zA-Z0-9]+", |lex| lex.slice(), priority = 2)]
    Unknown(&'hdr str),

    #[regex(r"\s+", |lex| lex.slice(), priority = 3)]
    WhiteSpaces(&'hdr str),
}

pub fn parse_dkim_policy_property_key<'hdr>(
    lexer: &mut Lexer<'hdr, DkimPolicyPropertyKeyToken<'hdr>>,
) -> Result<DkimPolicyPropertyKey<'hdr>, AuthResultsError<'hdr>> {
    while let Some(token) = lexer.next() {
        match token {
            Ok(DkimPolicyPropertyKeyToken::Unknown(_)) => {
                let property = token.map_err(|_| AuthResultsError::ParsePtypeBugPropertyGating)?;
                let mapped_property_res: Result<
                    DkimPolicyPropertyKey<'hdr>,
                    AuthResultsError<'hdr>,
                > = property.try_into();
                let mapped_property = mapped_property_res
                    .map_err(|_| AuthResultsError::ParsePtypeBugInvalidProperty)?;
                return Ok(mapped_property);
            }
            Ok(DkimPolicyPropertyKeyToken::WhiteSpaces(_)) => {
                // cont
            }
            Ok(DkimPolicyPropertyKeyToken::CommentStart) => {
                let mut comment_lexer = CommentToken::lexer(lexer.remainder());
                match parse_comment(&mut comment_lexer) {
                    Ok(_comment) => {}
                    Err(e) => return Err(e),
                }
                lexer.bump(comment_lexer.span().end);
            }
            _ => {
                let cut_slice = &lexer.source()[lexer.span().start..];
                let cut_span = &lexer.source()[lexer.span().start..lexer.span().end];

                let detail = crate::error::ParsingDetail {
                    component: "parse_dkim_policy_property_key",
                    span_start: lexer.span().start,
                    span_end: lexer.span().end,
                    source: lexer.source(),
                    clipped_span: cut_span,
                    clipped_remaining: cut_slice,
                };
                return Err(AuthResultsError::ParsingDetailed(detail));
            }
        }
    }
    Err(AuthResultsError::RunAwayDkimPropertyKey)
}
