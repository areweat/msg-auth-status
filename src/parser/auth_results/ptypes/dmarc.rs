//! dmarc ptype and it's properties
//!
//! See IANA Assignments
//! https://www.iana.org/assignments/email-auth/email-auth.xhtml
//!
//! And Auth in RFC 8601 s. 2.7

#[derive(Debug)]
pub enum DmarcProperty<'hdr> {
    /// RFC 7489
    HeaderFrom(&'hdr str),
    /// RFC 7489
    PolicyDmarc(&'hdr str),
}

#[derive(Debug)]
pub enum DmarcPtype {
    Header,
    HeaderDotFrom,
    Policy,
    PolicyDotDmarc,
}
