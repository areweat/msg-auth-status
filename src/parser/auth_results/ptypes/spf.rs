//! Parsing spf property types & values
//! https://www.iana.org/assignments/email-auth/email-auth.xhtml

use super::*;

pub mod spf_property_key;
pub mod spf_property_value;
