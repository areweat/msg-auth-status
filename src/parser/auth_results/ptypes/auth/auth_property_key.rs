//! Parsing auth property types & values
//! https://www.iana.org/assignments/email-auth/email-auth.xhtml

use crate::auth::*;

use super::ResultCodeError;

/// IANA Email Authentication Methods ptype / property Mapping stages
#[derive(Debug, PartialEq)]
pub enum AuthSmtpPropertyKey {
    MailFrom,
    Auth,
}

impl<'hdr> TryFrom<AuthSmtpPropertyKeyToken<'hdr>> for AuthSmtpPropertyKey {
    type Error = ResultCodeError;
    fn try_from(token: AuthSmtpPropertyKeyToken<'hdr>) -> Result<Self, Self::Error> {
        let okk = match token {
            AuthSmtpPropertyKeyToken::MailFrom => Self::MailFrom,
            AuthSmtpPropertyKeyToken::Auth => Self::Auth,
            _ => return Err(ResultCodeError::ParsePtypeBugInvalidProperty),
        };
        Ok(okk)
    }
}

//----------
// Parsing auth property
// https://www.iana.org/assignments/email-auth/email-auth.xhtml
//----------

use super::{parse_comment, CommentToken};

use logos::{Lexer, Logos};

#[derive(Debug, Logos)]
pub enum AuthSmtpPropertyKeyToken<'hdr> {
    #[token("mailfrom", priority = 1)]
    MailFrom,

    #[token("auth", priority = 1)]
    Auth,

    #[token("(", priority = 2)]
    CommentStart,

    #[regex(r"\s+", |lex| lex.slice(), priority = 6)]
    WhiteSpaces(&'hdr str),
}

pub fn parse_auth_smtp_property_key<'hdr>(
    lexer: &mut Lexer<'hdr, AuthSmtpPropertyKeyToken<'hdr>>,
) -> Result<AuthSmtpPropertyKey, ResultCodeError> {
    while let Some(token) = lexer.next() {
        match token {
            Ok(AuthSmtpPropertyKeyToken::MailFrom | AuthSmtpPropertyKeyToken::Auth) => {
                let property = token.map_err(|_| ResultCodeError::ParsePtypeBugPropertyGating)?;
                let mapped_property_res: Result<AuthSmtpPropertyKey, ResultCodeError> =
                    property.try_into();
                let mapped_property = mapped_property_res
                    .map_err(|_| ResultCodeError::ParsePtypeBugInvalidProperty)?;
                return Ok(mapped_property);
            }
            Ok(AuthSmtpPropertyKeyToken::WhiteSpaces(_)) => {
                // cont
            }
            Ok(AuthSmtpPropertyKeyToken::CommentStart) => {
                let mut comment_lexer = CommentToken::lexer(lexer.remainder());
                match parse_comment(&mut comment_lexer) {
                    Ok(comment) => {}
                    Err(e) => return Err(e),
                }
                lexer.bump(comment_lexer.span().end);
                //*lexer = X::lexer(comment_lexer.remainder());
            }
            _ => {
                let cut_slice = &lexer.source()[lexer.span().start..];
                let cut_span = &lexer.source()[lexer.span().start..lexer.span().end];

                panic!(
                    "parse_auth_property_key -- Invalid token {:?} - span = {:?}\n - Source = {:?}\n - Clipped/span: {:?}\n - Clipped/remaining: {:?}",
                    token,
                    lexer.span(),
                    lexer.source(),
                    cut_span,
	                cut_slice,
                );
            }
        }
    }
    Err(ResultCodeError::RunAwayAuthPropertyKey)
}
