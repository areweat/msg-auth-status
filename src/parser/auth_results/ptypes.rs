//! Property types Parser

use crate::dkim::DkimProperty;

/*
use crate::auth::AuthProperty;
use crate::dmarc::DmarcProperty;
use crate;:iprev::IpRevProperty;
use crate::spf::SpfProperty;
*/

// Parse into this public type
use crate::Prop;

use super::ResultCodeError;
use super::{parse_comment, CommentToken};
use super::{parse_reason, ReasonToken};
use super::{ParseCurrentResultChoice, ParseCurrentResultCode};
use logos::{Lexer, Logos};

mod auth;

mod dkim;
use dkim::dkim_property_key::{
    parse_dkim_header_property_key, DkimHeaderPropertyKey, DkimHeaderPropertyKeyToken,
};
use dkim::dkim_property_value::{parse_dkim_header_property_value, DkimHeaderPropertyValueToken};

mod dmarc;
mod iprev;
//mod policy;
mod spf;

#[derive(Debug, Default, PartialEq)]
pub enum PropTypeKey {
    #[default]
    Nothing,
    //Auth(AuthPtype),
    DkimHeader(DkimHeaderPropertyKey),
    //Dmarc(DmarcPtype),
    //IpRev(IpRevPtype),
    //Spf(SpfPtype),
}

#[derive(Debug, Default, PartialEq)]
pub enum PtypeChoice {
    #[default]
    Nothing,
    AuthSmtp,
    DkimHeader,
    DkimPolicy, // not sure
    IpRevPolicy,
    SpfSmtp,
}

impl PtypeChoice {
    fn from_associated_method_ptype<'hdr>(
        cur_res: &'hdr Option<ParseCurrentResultChoice<'hdr>>,
        token: &'hdr PtypeToken<'hdr>,
    ) -> Self {
        match cur_res {
            Some(ParseCurrentResultChoice::Dkim(_)) => match token {
                PtypeToken::PtypeHeader => Self::DkimHeader,
                PtypeToken::PtypePolicy => Self::DkimPolicy,
                _ => Self::Nothing,
            },
            Some(ParseCurrentResultChoice::Spf(_)) => match token {
                PtypeToken::PtypeSmtp => Self::SpfSmtp,
                _ => Self::Nothing,
            },
            Some(ParseCurrentResultChoice::SmtpAuth(_)) => match token {
                PtypeToken::PtypeSmtp => Self::AuthSmtp,
                _ => Self::Nothing,
            },
            Some(ParseCurrentResultChoice::IpRev(_)) => match token {
                PtypeToken::PtypePolicy => Self::IpRevPolicy,
                _ => Self::Nothing,
            },
            _ => Self::Nothing,
        }
    }
}

#[derive(Debug, Logos)]
pub enum PtypeToken<'hdr> {
    #[token("header", priority = 1)]
    PtypeHeader,

    #[token("smtp", priority = 1)]
    PtypeSmtp,

    #[token("policy", priority = 1)]
    PtypePolicy,

    #[token("=", priority = 2)]
    Equal,

    #[token(".", priority = 2)]
    Dot,

    #[token(";", priority = 3)]
    FieldSep,

    #[token("(", priority = 3)]
    CommentStart,

    #[token("reason", priority = 5)]
    Reason,

    #[regex(r"\s+", |lex| lex.slice(), priority = 6)]
    WhiteSpaces(&'hdr str),
}

#[derive(Debug, PartialEq)]
enum PtypeStage {
    WantPtype,
    GotPtype,
    WantDot,
    WantPropertyKey,
    GotPropertyKey,
    WantReasonEq,
    WantEq,
    GotEq,
    WantPropertyVal,
    GotPropertyVal,
}

impl PtypeStage {
    fn should_ignore_whitespace(&self) -> bool {
        match self {
            _ => true,
        }
    }
}

pub fn parse_ptype_properties<'hdr>(
    lexer: &mut Lexer<'hdr, PtypeToken<'hdr>>,
    cur_res: &mut Option<ParseCurrentResultChoice<'hdr>>,
) -> Result<(), ResultCodeError> {
    let mut stage = PtypeStage::WantPtype;
    let mut cur_ptype: PtypeChoice = PtypeChoice::Nothing;
    let mut cur_property: PropTypeKey = PropTypeKey::Nothing;

    let mut cur_res_choice = match &cur_res {
        Some(choice) => choice,
        None => return Err(ResultCodeError::ParsePtypeNoMethodResult),
    };

    while let Some(token) = lexer.next() {
        match token {
            Ok(PtypeToken::CommentStart) => {
                let mut comment_lexer = CommentToken::lexer(lexer.remainder());
                match parse_comment(&mut comment_lexer) {
                    Ok(comment) => {}
                    Err(e) => return Err(e),
                }
                lexer.bump(comment_lexer.span().end);
            }
            Ok(PtypeToken::PtypeSmtp | PtypeToken::PtypeHeader | PtypeToken::PtypePolicy)
                if stage == PtypeStage::WantPtype =>
            {
                let token_unwrap = token.expect("BUG: Incorrect gating.");
                let cur_ptype_try =
                    PtypeChoice::from_associated_method_ptype(cur_res, &token_unwrap);

                match cur_ptype_try {
                    PtypeChoice::Nothing => {
                        return Err(ResultCodeError::ParsePtypeBugInvalidProperty)
                    }
                    _ => {
                        cur_ptype = cur_ptype_try;
                    }
                }
                stage = PtypeStage::WantDot;
            }
            Ok(PtypeToken::Dot)
                if stage == PtypeStage::WantDot && cur_ptype != PtypeChoice::Nothing =>
            {
                cur_property = match cur_ptype {
                    PtypeChoice::DkimHeader => {
                        let mut property_key_lexer =
                            DkimHeaderPropertyKeyToken::lexer(lexer.remainder());
                        let property_key =
                            match parse_dkim_header_property_key(&mut property_key_lexer) {
                                Err(e) => return Err(e),
                                Ok(property_key) => property_key,
                            };
                        lexer.bump(property_key_lexer.span().end);
                        PropTypeKey::DkimHeader(property_key)
                    }
                    _ => return Err(ResultCodeError::PropertiesNotImplemented),
                };
                stage = PtypeStage::WantEq;
            }

            Ok(PtypeToken::FieldSep) => {
                break;
            }
            Ok(PtypeToken::WhiteSpaces(ref wsh)) if stage.should_ignore_whitespace() => {
                // cont
            }
            Ok(PtypeToken::Reason) if stage == PtypeStage::WantPtype => {
                stage = PtypeStage::WantReasonEq;
            }
            Ok(PtypeToken::Equal) if stage == PtypeStage::WantReasonEq => {
                let mut reason_lexer = ReasonToken::lexer(lexer.remainder());
                let reason_res = match parse_reason(&mut reason_lexer) {
                    Err(e) => return Err(e),
                    Ok(reason) => reason,
                };
                match cur_res {
                    Some(ref mut ref_choice) => {
                        ref_choice.set_reason(reason_res);
                    }
                    _ => {}
                }
                lexer.bump(reason_lexer.span().end);
                stage = PtypeStage::WantPtype;
            }
            Ok(PtypeToken::Equal)
                if stage == PtypeStage::WantEq && cur_property != PropTypeKey::Nothing =>
            {
                match cur_property {
                    PropTypeKey::DkimHeader(ref property) => {
                        let mut property_value_lexer =
                            DkimHeaderPropertyValueToken::lexer(lexer.remainder());
                        let property_value = match parse_dkim_header_property_value(
                            &mut property_value_lexer,
                            &property,
                            //&mut cur_res,
                        ) {
                            Err(e) => return Err(e),
                            Ok(property_value) => property_value,
                        };
                        lexer.bump(property_value_lexer.span().end);

                        match cur_res {
                            Some(ParseCurrentResultChoice::Dkim(ref mut dkim_res)) => {
                                dkim_res.set_header(&property_value);
                            }
                            _ => {}
                        }
                    }
                    _ => {
                        return Err(ResultCodeError::PropertyValuesNotImplemented);
                    }
                };
            }
            _ => {
                let cut_slice = &lexer.source()[lexer.span().start..];
                let cut_span = &lexer.source()[lexer.span().start..lexer.span().end];
                panic!(
                    "parse_ptypes_properties({:?}) -- Invalid token {:?} - span = {:?}\n - Source = {:?}\n - Clipped/span: {:?}\n - Clipped/remaining: {:?}",
                    stage,
                    token,
                    lexer.span(),
                    lexer.source(),
                    cut_span,
                    cut_slice,
                );
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod test {

    use super::*;
    use insta::assert_debug_snapshot;
    use rstest::rstest;

    use crate::spf::*;

    fn prep_spf<'hdr>(s: &'hdr str) -> ParseCurrentResultCode<'hdr> {
        let mut current = ParseCurrentResultCode::default();
        let mut spf_result = SpfResult::default();
        current.result = Some(ParseCurrentResultChoice::Spf(spf_result));
        current
    }

    #[rstest]
    #[case("smtp.mailfrom=example.net", 1)]
    fn spf(#[case] prop_str: &str, #[case] expected_count: usize) {
        let mut cur_spf = prep_spf(prop_str);
        let mut lexer = PtypeToken::lexer(prop_str);

        let res = parse_ptype_properties(&mut lexer, &mut cur_spf.result).unwrap();

        assert_debug_snapshot!(cur_spf);
        //assert_eq!(expected_count, res.properties.len());
    }
}
