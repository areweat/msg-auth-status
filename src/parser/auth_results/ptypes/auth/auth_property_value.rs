//! Parsing auth property values

use crate::auth::ptypes::AuthSmtp;
use crate::error::AuthResultsError;

use super::AuthSmtpPropertyKey;
use super::{parse_comment, CommentToken};

use logos::{Lexer, Logos};

#[derive(Debug, Logos)]
pub enum AuthSmtpPropertyValueToken<'hdr> {
    #[token("(", priority = 1)]
    CommentStart,

    #[regex(r#"[^(\s\r\n\t;]+"#, |lex| lex.slice(), priority = 2)]
    MaybeValue(&'hdr str),

    #[regex(r"[\s\r\n\t]+", |lex| lex.slice(), priority = 3)]
    Whs(&'hdr str),
}

impl<'hdr> AuthSmtp<'hdr> {
    fn from_parsed(pkey: &AuthSmtpPropertyKey, val: &'hdr str) -> Self {
        match pkey {
            AuthSmtpPropertyKey::MailFrom => AuthSmtp::MailFrom(val),
            AuthSmtpPropertyKey::Auth => AuthSmtp::Auth(val),
        }
    }
}

pub fn parse_auth_smtp_property_value<'hdr>(
    lexer: &mut Lexer<'hdr, AuthSmtpPropertyValueToken<'hdr>>,
    property_key: &AuthSmtpPropertyKey,
) -> Result<AuthSmtp<'hdr>, AuthResultsError<'hdr>> {
    let mut cur_res: Option<AuthSmtp<'hdr>> = None;

    while let Some(token) = lexer.next() {
        match token {
            Ok(AuthSmtpPropertyValueToken::MaybeValue(val)) => {
                cur_res = Some(AuthSmtp::from_parsed(property_key, val));
                break;
            }
            Ok(AuthSmtpPropertyValueToken::CommentStart) => {
                let mut comment_lexer = CommentToken::lexer(lexer.remainder());
                match parse_comment(&mut comment_lexer) {
                    Ok(_comment) => {}
                    Err(e) => return Err(e),
                }
                lexer.bump(comment_lexer.span().end);
            }
            Ok(AuthSmtpPropertyValueToken::Whs(_)) => {
                // cont
            }
            _ => {
                let cut_slice = &lexer.source()[lexer.span().start..];
                let cut_span = &lexer.source()[lexer.span().start..lexer.span().end];

                let detail = crate::error::ParsingDetail {
                    component: "auth_property_value",
                    span_start: lexer.span().start,
                    span_end: lexer.span().end,
                    source: lexer.source(),
                    clipped_span: cut_span,
                    clipped_remaining: cut_slice,
                };
                return Err(AuthResultsError::ParsingDetailed(detail));
            }
        }
    }

    if let Some(value) = cur_res {
        return Ok(value);
    }

    Err(AuthResultsError::RunAwayAuthPropertyValue)
}
