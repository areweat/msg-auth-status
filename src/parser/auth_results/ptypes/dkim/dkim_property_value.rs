//! Parsing dkim property values

use crate::dkim::ptypes::DkimHeader;
use crate::dkim::ptypes::DkimPolicy;
use crate::dkim::*;

use super::ResultCodeError;

use super::DkimHeaderPropertyKey;
use super::DkimPolicyPropertyKey;
use super::{parse_comment, CommentToken};

use logos::{Lexer, Logos};

//---------------------------------------
// DKIM header.xxx ptype values
//---------------------------------------

#[derive(Debug, Logos)]
pub enum DkimHeaderPropertyValueToken<'hdr> {
    #[token("(", priority = 1)]
    CommentStart,

    #[regex(r#"[^\s;]+"#, |lex| lex.slice(), priority = 2)]
    MaybeValue(&'hdr str),

    #[regex(r"\s+", |lex| lex.slice(), priority = 3)]
    WhiteSpaces(&'hdr str),
}

impl<'hdr> DkimHeader<'hdr> {
    fn from_parsed(pkey: &DkimHeaderPropertyKey, val: &'hdr str) -> Self {
        match pkey {
            DkimHeaderPropertyKey::TagD => DkimHeader::D(val),
            DkimHeaderPropertyKey::TagI => DkimHeader::I(val),
            DkimHeaderPropertyKey::TagB => DkimHeader::B(val),
            DkimHeaderPropertyKey::TagA => {
                let alg = match val {
                    "rsa-sha1" => DkimAlgorithm::Rsa_Sha1,
                    "rsa-sha256" => DkimAlgorithm::Rsa_Sha256,
                    "ed25519-sha256" => DkimAlgorithm::Ed25519_Sha256,
                    _ => DkimAlgorithm::Unknown(val),
                };
                DkimHeader::A(alg)
            }
            DkimHeaderPropertyKey::TagS => DkimHeader::S(val),
            DkimHeaderPropertyKey::Rfc5322From => DkimHeader::Rfc5322From(val), // not covered
        }
    }
}

pub fn parse_dkim_header_property_value<'hdr>(
    lexer: &mut Lexer<'hdr, DkimHeaderPropertyValueToken<'hdr>>,
    property_key: &DkimHeaderPropertyKey,
) -> Result<DkimHeader<'hdr>, ResultCodeError> {
    let mut value_captured = false;
    let mut cur_res: Option<DkimHeader<'hdr>> = None;

    while let Some(token) = lexer.next() {
        match token {
            Ok(DkimHeaderPropertyValueToken::MaybeValue(val)) if value_captured == false => {
                cur_res = Some(DkimHeader::from_parsed(property_key, val));
                value_captured = true;
                break;
            }
            Ok(DkimHeaderPropertyValueToken::CommentStart) => {
                let mut comment_lexer = CommentToken::lexer(lexer.remainder());
                match parse_comment(&mut comment_lexer) {
                    Ok(comment) => {}
                    Err(e) => return Err(e),
                }
                lexer.bump(comment_lexer.span().end);
                //*lexer = X::lexer(comment_lexer.remainder());
            }
            Ok(DkimHeaderPropertyValueToken::WhiteSpaces(_)) => {
                // cont
            }
            _ => {
                let cut_slice = &lexer.source()[lexer.span().start..];
                let cut_span = &lexer.source()[lexer.span().start..lexer.span().end];

                panic!(
                    "parse_dkim_header_property_value -- Invalid token {:?} - span = {:?}\n - Source = {:?}\n - Clipped/span: {:?}\n - Clipped/remaining: {:?}",
                    token,
                    lexer.span(),
                    lexer.source(),
                    cut_span,
	                cut_slice,
                );
            }
        }
    }

    if let Some(value) = cur_res {
        return Ok(value);
    }

    Err(ResultCodeError::RunAwayDkimPropertyKey)
}

//---------------------------------------
// DKIM policy.xxx ptype values
//---------------------------------------

#[derive(Debug, Logos)]
pub enum DkimPolicyPropertyValueToken<'hdr> {
    #[token("(", priority = 1)]
    CommentStart,

    #[regex(r"\s+", |lex| lex.slice(), priority = 2)]
    WhiteSpaces(&'hdr str),

    #[regex(r"\n+", |lex| lex.slice(), priority = 3)]
    LineFeeds(&'hdr str),

    #[regex(r"[A-Za-z0-9]+", |lex| lex.slice(), priority = 4)]
    MaybeValue(&'hdr str),
}

impl<'hdr> DkimPolicy<'hdr> {
    fn from_parsed(pkey: &DkimPolicyPropertyKey<'hdr>, val: &'hdr str) -> Self {
        match pkey {
            DkimPolicyPropertyKey::Unknown(key) => DkimPolicy::Unknown(key, val),
        }
    }
}

pub fn parse_dkim_policy_property_value<'hdr>(
    lexer: &mut Lexer<'hdr, DkimPolicyPropertyValueToken<'hdr>>,
    property_key: &DkimPolicyPropertyKey<'hdr>,
) -> Result<DkimPolicy<'hdr>, ResultCodeError> {
    let mut value_captured = false;
    let mut cur_res: Option<DkimPolicy<'hdr>> = None;

    while let Some(token) = lexer.next() {
        match token {
            Ok(DkimPolicyPropertyValueToken::MaybeValue(val)) if value_captured == false => {
                cur_res = Some(DkimPolicy::from_parsed(property_key, val));
                value_captured = true;
                break;
            }
            Ok(DkimPolicyPropertyValueToken::CommentStart) => {
                let mut comment_lexer = CommentToken::lexer(lexer.remainder());
                let comment = match parse_comment(&mut comment_lexer) {
                    Ok(comment) => comment,
                    Err(e) => return Err(e),
                };
                lexer.bump(comment_lexer.span().end);
            }
            Ok(DkimPolicyPropertyValueToken::WhiteSpaces(_)) => {
                // cont
            }
            _ => {
                let cut_slice = &lexer.source()[lexer.span().start..];
                let cut_span = &lexer.source()[lexer.span().start..lexer.span().end];

                panic!(
                    "parse_dkim_policy_property_value -- Invalid token {:?} - span = {:?}\n - Source = {:?}\n - Clipped/span: {:?}\n - Clipped/remaining: {:?}",
                    token,
                    lexer.span(),
                    lexer.source(),
                    cut_span,
	                cut_slice,
                );
            }
        }
    }

    if let Some(value) = cur_res {
        return Ok(value);
    }

    Err(ResultCodeError::RunAwayDkimPropertyKey)
}
