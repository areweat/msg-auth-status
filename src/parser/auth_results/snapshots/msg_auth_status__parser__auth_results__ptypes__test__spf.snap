---
source: src/parser/auth_results/ptypes.rs
expression: cur_spf
---
ParseCurrentResultCode {
    result: Some(
        Spf(
            SpfResult {
                code: Unknown,
                reason: None,
                smtp_mailfrom: Some(
                    "example.net",
                ),
                smtp_helo: None,
                raw: None,
            },
        ),
    ),
    raw: None,
}
